File: innerlibs1/JSminifier.vb
Imports System.IO

Public Class JSMin
    Const EOF As Integer = -1
    Private sr As StringReader
    Private sw As String = ""
    Private theA As Integer
    Private theB As Integer
    Private theLookahead As Integer = EOF

    Public Function Minify(ByVal src As String) As String

        Using sr = New StringReader(src)
            jsmin(sr)
            Return sw
        End Using

    End Function

    Private Sub jsmin(sr As StringReader)
        theA = AscW(vbLf)
        action(3, sr)

        While theA <> EOF

            Select Case theA
                Case AscW(" ")

                    If IsAlphanum(theB) Then
                        action(1, sr)
                    Else
                        action(2, sr)
                    End If

                    Exit Select
                Case AscW(vbLf)

                    Select Case theB
                        Case AscW("{"), AscW("["), AscW("("), AscW("+"), AscW("-")
                            action(1, sr)
                            Exit Select
                        Case AscW(" ")
                            action(3, sr)
                            Exit Select
                        Case Else

                            If IsAlphanum(theB) Then
                                action(1, sr)
                            Else
                                action(2, sr)
                            End If

                            Exit Select
                    End Select

                    Exit Select
                Case Else

                    Select Case theB
                        Case AscW(" ")

                            If IsAlphanum(theA) Then
                                action(1, sr)
                                Exit Select
                            End If

                            action(3, sr)
                            Exit Select
                        Case AscW(vbLf)

                            Select Case theA
                                Case AscW("}"), AscW("]"), AscW(")"), AscW("+"), AscW("-"), AscW(""""), AscW("'")
                                    action(1, sr)
                                    Exit Select
                                Case Else

                                    If IsAlphanum(theA) Then
                                        action(1, sr)
                                    Else
                                        action(3, sr)
                                    End If

                                    Exit Select
                            End Select

                            Exit Select
                        Case Else
                            action(1, sr)
                            Exit Select
                    End Select

                    Exit Select
            End Select
        End While
    End Sub

    Private Sub action(ByVal d As Integer, sr As StringReader)
        If d <= 1 Then
            put(theA)
        End If

        If d <= 2 Then
            theA = theB

            If theA = AscW("'") OrElse theA = AscW("""") Then

                While True
                    put(theA)
                    theA = [get](sr)

                    If theA = theB Then
                        Exit While
                    End If

                    If theA <= AscW(vbLf) Then
                        Throw New Exception(String.Format("Error: JSMIN unterminated string literal: {0}" & vbLf, theA))
                    End If

                    If theA = AscW("\") Then
                        put(theA)
                        theA = [get](sr)
                    End If
                End While
            End If
        End If

        If d <= 3 Then
            theB = [next](sr)

            If theB = AscW("/") AndAlso (theA = AscW("(") OrElse theA = AscW(",") OrElse theA = AscW("=") OrElse theA = AscW("[") OrElse theA = AscW("!") OrElse theA = AscW(":") OrElse theA = AscW("&") OrElse theA = AscW("|") OrElse theA = AscW("?") OrElse theA = AscW("{") OrElse theA = AscW("}") OrElse theA = AscW(";") OrElse theA = AscW(vbLf)) Then
                put(theA)
                put(theB)

                While True
                    theA = [get](sr)

                    If theA = AscW("/") Then
                        Exit While
                    ElseIf theA = AscW("\") Then
                        put(theA)
                        theA = [get](sr)
                    ElseIf theA <= AscW(vbLf) Then
                        Throw New Exception(String.Format("Error: JSMIN unterminated Regular Expression literal : {0}." & vbLf, theA))
                    End If

                    put(theA)
                End While

                theB = [next](sr)
            End If
        End If
    End Sub

    Private Function [next](sr As StringReader) As Integer
        Dim c As Integer = [get](sr)

        If c = AscW("/") Then

            Select Case peek(sr)
                Case AscW("/")

                    While True
                        c = [get](sr)

                        If c <= AscW(vbLf) Then
                            Return c
                        End If
                    End While

                Case AscW("*")
                    [get](sr)

                    While True

                        Select Case [get](sr)
                            Case AscW("*")

                                If peek(sr) = AscW("/") Then
                                    [get](sr)
                                    Return AscW(" ")
                                End If

                                Exit Select
                            Case EOF
                                Throw New Exception("Error: JSMIN Unterminated comment." & vbLf)
                        End Select
                    End While

                Case Else
                    Return c
            End Select
        End If

        Return c
    End Function

    Private Function peek(sr As StringReader) As Integer
        theLookahead = [get](sr)
        Return theLookahead
    End Function

    Private Function [get](sr As StringReader) As Integer
        Dim c As Integer = theLookahead
        theLookahead = EOF

        If c = EOF Then
            c = sr.Read()
        End If

        If c >= AscW(" ") OrElse c = AscW(vbLf) OrElse c = EOF Then
            Return c
        End If

        If c = AscW(vbCr) Then
            Return AscW(vbLf)
        End If

        Return AscW(" ")
    End Function

    Private Sub put(ByVal c As Integer)
        sw = sw.Append(ChrW(c))
    End Sub

    Function IsAlphanum(ByVal c As Integer) As Boolean
        Return ((c >= AscW("a") AndAlso c <= AscW("z")) OrElse (c >= AscW("0") AndAlso c <= AscW("9")) OrElse (c >= AscW("A") AndAlso c <= AscW("Z")) OrElse c = AscW("_") OrElse c = AscW("$") OrElse c = AscW("\") OrElse c = AscW("+") OrElse c > 126)
    End Function

End Class

File: InnerLibs/JSminifier.cs
using System;
using System.IO;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace InnerLibs
{
    public class JSMin
    {
        private const int EOF = -1;
        private StringReader sr;
        private string sw = "";
        private int theA;
        private int theB;
        private int theLookahead = EOF;

        public string Minify(string src)
        {
            using (var sr = new StringReader(src))
            {
                jsmin(sr);
                return sw;
            }
        }

        private void jsmin(StringReader sr)
        {
            theA = Strings.AscW(Constants.vbLf);
            action(3, sr);
            while (theA != EOF)
            {
                switch (theA)
                {
                    case var @case when @case == Strings.AscW(" "):
                        {
                            if (IsAlphanum(theB))
                            {
                                action(1, sr);
                            }
                            else
                            {
                                action(2, sr);
                            }

                            break;
                        }

                    case var case1 when case1 == Strings.AscW(Constants.vbLf):
                        {
                            switch (theB)
                            {
                                case var case2 when case2 == Strings.AscW("{"):
                                case var case3 when case3 == Strings.AscW("["):
                                case var case4 when case4 == Strings.AscW("("):
                                case var case5 when case5 == Strings.AscW("+"):
                                case var case6 when case6 == Strings.AscW("-"):
                                    {
                                        action(1, sr);
                                        break;
                                    }

                                case var case7 when case7 == Strings.AscW(" "):
                                    {
                                        action(3, sr);
                                        break;
                                    }

                                default:
                                    {
                                        if (IsAlphanum(theB))
                                        {
                                            action(1, sr);
                                        }
                                        else
                                        {
                                            action(2, sr);
                                        }

                                        break;
                                    }
                            }

                            break;
                        }

                    default:
                        {
                            switch (theB)
                            {
                                case var case8 when case8 == Strings.AscW(" "):
                                    {
                                        if (IsAlphanum(theA))
                                        {
                                            action(1, sr);
                                            break;
                                        }

                                        action(3, sr);
                                        break;
                                    }

                                case var case9 when case9 == Strings.AscW(Constants.vbLf):
                                    {
                                        switch (theA)
                                        {
                                            case var case10 when case10 == Strings.AscW("}"):
                                            case var case11 when case11 == Strings.AscW("]"):
                                            case var case12 when case12 == Strings.AscW(")"):
                                            case var case13 when case13 == Strings.AscW("+"):
                                            case var case14 when case14 == Strings.AscW("-"):
                                            case var case15 when case15 == Strings.AscW("\""):
                                            case var case16 when case16 == Strings.AscW("'"):
                                                {
                                                    action(1, sr);
                                                    break;
                                                }

                                            default:
                                                {
                                                    if (IsAlphanum(theA))
                                                    {
                                                        action(1, sr);
                                                    }
                                                    else
                                                    {
                                                        action(3, sr);
                                                    }

                                                    break;
                                                }
                                        }

                                        break;
                                    }

                                default:
                                    {
                                        action(1, sr);
                                        break;
                                    }
                            }

                            break;
                        }
                }
            }
        }

        private void action(int d, StringReader sr)
        {
            if (d <= 1)
            {
                put(theA);
            }

            if (d <= 2)
            {
                theA = theB;
                if (theA == Strings.AscW("'") || theA == Strings.AscW("\""))
                {
                    while (true)
                    {
                        put(theA);
                        theA = get(sr);
                        if (theA == theB)
                        {
                            break;
                        }

                        if (theA <= Strings.AscW(Constants.vbLf))
                        {
                            throw new Exception(string.Format("Error: JSMIN unterminated string literal: {0}" + Constants.vbLf, theA));
                        }

                        if (theA == Strings.AscW(@"\"))
                        {
                            put(theA);
                            theA = get(sr);
                        }
                    }
                }
            }

            if (d <= 3)
            {
                theB = next(sr);
                if (theB == Strings.AscW("/") && (theA == Strings.AscW("(") || theA == Strings.AscW(",") || theA == Strings.AscW("=") || theA == Strings.AscW("[") || theA == Strings.AscW("!") || theA == Strings.AscW(":") || theA == Strings.AscW("&") || theA == Strings.AscW("|") || theA == Strings.AscW("?") || theA == Strings.AscW("{") || theA == Strings.AscW("}") || theA == Strings.AscW(";") || theA == Strings.AscW(Constants.vbLf)))
                {
                    put(theA);
                    put(theB);
                    while (true)
                    {
                        theA = get(sr);
                        if (theA == Strings.AscW("/"))
                        {
                            break;
                        }
                        else if (theA == Strings.AscW(@"\"))
                        {
                            put(theA);
                            theA = get(sr);
                        }
                        else if (theA <= Strings.AscW(Constants.vbLf))
                        {
                            throw new Exception(string.Format("Error: JSMIN unterminated Regular Expression literal : {0}." + Constants.vbLf, theA));
                        }

                        put(theA);
                    }

                    theB = next(sr);
                }
            }
        }

        private int next(StringReader sr)
        {
            int c = get(sr);
            if (c == Strings.AscW("/"))
            {
                switch (peek(sr))
                {
                    case var @case when @case == Strings.AscW("/"):
                        {
                            while (true)
                            {
                                c = get(sr);
                                if (c <= Strings.AscW(Constants.vbLf))
                                {
                                    return c;
                                }
                            }

                            break;
                        }

                    case var case1 when case1 == Strings.AscW("*"):
                        {
                            get(sr);
                            while (true)
                            {
                                switch (get(sr))
                                {
                                    case var case2 when case2 == Strings.AscW("*"):
                                        {
                                            if (peek(sr) == Strings.AscW("/"))
                                            {
                                                get(sr);
                                                return Strings.AscW(" ");
                                            }

                                            break;
                                        }

                                    case EOF:
                                        {
                                            throw new Exception("Error: JSMIN Unterminated comment." + Constants.vbLf);
                                            break;
                                        }
                                }
                            }

                            break;
                        }

                    default:
                        {
                            return c;
                        }
                }
            }

            return c;
        }

        private int peek(StringReader sr)
        {
            theLookahead = get(sr);
            return theLookahead;
        }

        private int get(StringReader sr)
        {
            int c = theLookahead;
            theLookahead = EOF;
            if (c == EOF)
            {
                c = sr.Read();
            }

            if (c >= Strings.AscW(" ") || c == Strings.AscW(Constants.vbLf) || c == EOF)
            {
                return c;
            }

            if (c == Strings.AscW(Constants.vbCr))
            {
                return Strings.AscW(Constants.vbLf);
            }

            return Strings.AscW(" ");
        }

        private void put(int c)
        {
            sw += Conversions.ToString(c);
        }

        public bool IsAlphanum(int c)
        {
            return c >= Strings.AscW("a") && c <= Strings.AscW("z") || c >= Strings.AscW("0") && c <= Strings.AscW("9") || c >= Strings.AscW("A") && c <= Strings.AscW("Z") || c == Strings.AscW("_") || c == Strings.AscW("$") || c == Strings.AscW(@"\") || c == Strings.AscW("+") || c > 126;
        }
    }
}